src/pages/signup/signup.test.js

```diff
@@ -0,0 +1,149 @@
+    target: {
+      value,
+    },
+  });
+};
+
+const fireSubmit = () => {
+  fireEvent.click(screen.getByRole('button', { name: 'Sign up' }));
+};
+
+const testCases = (type, dataCases) => {
+  const title = `Render ${type} $field error message`;
+  test.each(dataCases)(title, async ({ field, data, error }) => {
+    render(<SignUp />);
+
+    const inpField = screen.getByLabelText(field);
+    const inpData = data;
+
+    handleInput(inpField, inpData);
+    fireSubmit();
+
+    await waitFor(() => {
+      expect(screen.getByText(error)).toBeInTheDocument();
+    });
+  });
+};
+
+describe('Sign up functionality', () => {
+  beforeAll(() => {
+    localStorage.setItem('users', JSON.stringify(userData));
+  });
+
+  const invalidCases = [
+    {
+      field: 'email',
+      data: 'toryemail',
+      error: 'Email is invalid',
+    },
+    {
+      field: 'username',
+      data: 'tory',
+      error:
+        'Username must be between 6-20 length with no special character',
+    },
+    {
+      field: 'phone',
+      data: '012345678',
+      error: 'Phone number must be 10-digits',
+    },
+    {
+      field: 'password1',
+      data: 'pass',
+      error: 'Password must be at least 6 characters',
+    },
+  ];
+
+  const existingCases = [
+    {
+      field: 'email',
+      data: 'quantriminh@gmail.com',
+      error: 'Email is already taken',
+    },
+    {
+      field: 'username',
+      data: 'minhtri',
+      error: 'Username is already taken',
+    },
+  ];
+
+  describe('Invalid fields', () => {
+    testCases('invalid', invalidCases);
+
+    test("Render passwords don't match error message", async () => {
+      render(<SignUp />);
+
+      const inpPW1 = screen.getByLabelText('password1');
+      const inpPW2 = screen.getByLabelText('password2');
+
+      handleInput(inpPW1, 'tory123');
+      handleInput(inpPW2, 'tory124');
+      fireSubmit();
+
+      await waitFor(() => {
+        expect(screen.getByText("Passwords don't match")).toBeInTheDocument();
+      });
+    })
+  });
+
+  describe('Existing fields', () => {
+    testCases('existing', existingCases);
+  });
+
+  describe('Signup successfully', () => {
+    test('Return new account record in localStorage', async () => {
+      render(<SignUp />);
+
+      const newUser = {
+        email: 'justtory@gmail.com',
+        username: 'justtory',
+        phone: '0123456789',
+        password1: 'heheboi123',
+        password2: 'heheboi123',
+      };
+
+      const emailField = screen.getByLabelText('email');
+      const usernameField = screen.getByLabelText('username');
+      const phoneField = screen.getByLabelText('phone');
+      const pw1Field = screen.getByLabelText('password1');
+      const pw2Field = screen.getByLabelText('password2');
+
+      handleInput(emailField, newUser.email);
+      handleInput(usernameField, newUser.username);
+      handleInput(phoneField, newUser.phone);
+      handleInput(pw1Field, newUser.password1);
```

Comment:
We can put the getByLabelText (finding selectors) inside the handle input function, just like this
<img width="570" alt="Screen Shot 2022-07-09 at 22 47 44" src="https://user-images.githubusercontent.com/107382512/178112901-5207f7aa-732a-4cdf-94ff-bd2748aa50ec.png">

Reply:
Noted, I will update them accordingly